kornia/losses/perceptual.py

```diff
@@ -0,0 +1,304 @@
+    Return:
+        the computed loss.
+    """
+    if not isinstance(patch_1, torch.Tensor):
+        raise TypeError(f"patch_1 type is not a torch.Tensor. Got {type(patch_1)}")
+
+    if not len(patch_1.shape) == 4:
+        raise ValueError(f"Invalid input shape of patch_1, we expect BxCxHxW. Got: {patch_1.shape}")
+
+    if not isinstance(patch_2, torch.Tensor):
+        raise TypeError(f"patch_2 type is not a torch.Tensor. Got {type(patch_2)}")
+
+    if not len(patch_2.shape) == 4:
+        raise ValueError(f"Invalid input shape of patch_2, we expect BxCxHxW. Got: {patch_2.shape}")
+
+    if patch_1.shape != patch_2.shape:
+        raise ValueError(f'Expected patch_1 shape ({patch_1.shape}) to match patch_2 shape ({patch_2.shape}).')
+
+    if not isinstance(delta_hat_12, torch.Tensor):
+        raise TypeError(f"delta_hat_12 type is not a torch.Tensor. Got {type(delta_hat_12)}")
+
+    if not len(delta_hat_12.shape) == 3 or not delta_hat_12.shape[1] == 4 or not delta_hat_12.shape[2] == 2:
+        raise ValueError(f"Invalid input shape of delta_hat_12, we expect Bx4x2. Got: {delta_hat_12.shape}")
+
+    if not delta_hat_12.size(0) == patch_1.size(0):
+        raise ValueError(f'Expected delta_hat_12 batch_size ({delta_hat_12.size(0)}) to match patch_1 batch size '
+                         f'({patch_1.size(0)}).')
+
+    if not isinstance(delta_hat_21, torch.Tensor):
+        raise TypeError(f"delta_hat_21 type is not a torch.Tensor. Got {type(delta_hat_21)}")
+
+    if not len(delta_hat_21.shape) == 3 or not delta_hat_21.shape[1] == 4 or not delta_hat_21.shape[2] == 2:
+        raise ValueError(f"Invalid input shape of delta_hat_21, we expect Bx4x2. Got: {delta_hat_21.shape}")
+
+    if not delta_hat_21.size(0) == patch_1.size(0):
+        raise ValueError(f'Expected delta_hat_21 batch_size ({delta_hat_21.size(0)}) to match patch_1 batch size '
+                         f'({patch_1.size(0)}).')
+
+    if not isinstance(loss_network, nn.Module):
+        raise TypeError(f"loss_network type is not a str. Got {type(loss_network)}")
+
+    # Compute features of both patches
+    patch_1_f = loss_network(patch_1)
+    patch_2_f = loss_network(patch_2)
+
+    # Warp patch 1 with delta hat_12
+    patch_1_prime, h1 = _warp(patch_1, delta_hat=delta_hat_12)
+    patch_1_prime_f = loss_network(patch_1_prime)
+
+    # Warp patch 2 with delta hat_21
+    patch_2_prime, h2 = _warp(patch_2, delta_hat=delta_hat_21)
+    patch_2_prime_f = loss_network(patch_2_prime)
+
+    # Create and warp masks
+    patch_1_m = torch.ones_like(patch_1)
+    patch_2_m = torch.ones_like(patch_2)
+    patch_1_m_prime, _ = _warp(patch_1_m, delta_hat=delta_hat_12)
+    patch_2_m_prime, _ = _warp(patch_2_m, delta_hat=delta_hat_21)
+
+    # Mask size mismatch downsampling
+    _, _, f_h, _ = patch_1_prime_f.shape
+    downsample_factor = patch_1_m.shape[-1] // f_h
+    downsample_layer = torch.nn.AvgPool2d(kernel_size=downsample_factor, stride=downsample_factor, padding=0)
```

Comment:
use the functional instead: https://pytorch.org/docs/stable/generated/torch.nn.functional.avg_pool2d.html?highlight=avg_pool#torch.nn.functional.avg_pool2d

Reply:
Done! Just being curious, why is it better to use?